### ISSUE TITLE
is_file() return False when the artifacts' name contains "#"

### ISSUE THREAD

**Issue by nabice** [devopshq/artifactory#206]
Version:
dohq-artifactory       0.7.475

**Comment by fuzzmz** [comment]
@nabice, can you provide some more details on how to reproduce this issue? Does it happen when uploading a file to Artifactory, or downloading it, or changing properties, etc.

**Comment by nabice** [comment]
@fuzzmz 
|----rootDir
       |-------C#.xml

```
rootDir = ArtifactoryPath(AF_URL)
for child in rootDir:
    print(child.is_file())
```
Output:
False

======

Expect:
True

**Comment by fuzzmz** [comment]
Problem is that `#` gets treated as a [fragment identifier](https://en.wikipedia.org/wiki/Fragment_identifier) and the requests library is treating is as such.

This means that in the initial URL `http://artifactory/rootDir/C#.xml` you have a network resource (URI) named `http://artifactory/rootDir/C` followed by a fragment, `.xml`.

As such, requests will try to access the storage API for C `http://artifactory/api/storage/rootDir/C` instead of `http://artifactory/api/storage/rootDir/C%23.xml`, which will return a 404 not found and leads to C#.xml not being reported as a file.

Now, the solution would be to encode the pound sign, but I'm not sure if it's something we want to do lightly.

@allburov, want to weigh in?

**Comment by allburov** [comment]
@fuzzmz Nice investigation! 
It does not seem that we will break something if we add this, does it? Let's do it then (and properly test it)! :)

**Comment by fuzzmz** [comment]
Will take a closer look through the codebase next week.

For sure we can just fix it in [is_file](https://github.com/devopshq/artifactory/blob/c72efeb5651e8c665b155be909da3e9bd566aa00/artifactory.py#L761) but I'd rather see if updating [rest_get](https://github.com/devopshq/artifactory/blob/c72efeb5651e8c665b155be909da3e9bd566aa00/artifactory.py#L568) has any unintended side-effects. That or just applying the normalization to any pathobjs. This would have the advantage of future-proofing it for any new methods we want to add using that API.

**Comment by fuzzmz** [comment]
Sorry for the delay. Still have this on my to-do list but unfortunately didn't yet get around to it.

**Comment by beliaev-maksim** [comment]
@fuzzmz @allburov 
there are some other characters that should be decoded then: https://en.wikipedia.org/wiki/Percent-encoding#Reserved_characters

may be just do a check of URL in each of requests and if any, then send a warning and ask user to apply new function `url_decode` on URL if user sees an issue
and in this `url_decode` function just convert all symbols to % representation?

**Comment by beliaev-maksim** [comment]
looks like #245 is related

**Comment by nacc** [comment]
Yeah, good catch. It is definitely related. I submitted a PR that fixes the cases I hit, but fixing it properly in the pathobj would be ideal. I think the issue is that using pathlib-like manipulation for URLs requires encoding the components of the URL to ensure that special characters are not transmitted as literals.

**Comment by beliaev-maksim** [comment]
@allburov please close as addressed by #254

**Comment by allburov** [comment]
@beliaev-maksim Hi! 
I have added you to artifactory's team, now you're able to close issues yourself (if I added right roles)

**Comment by beliaev-maksim** [comment]
@allburov 
looks like I still do not have permissions to close

**Comment by beliaev-maksim** [comment]
addressed by #254
